File: lib/core/services/web3/dashboard_service.dart
/// GramPulse Dashboard Service - Transparency Dashboard Client
///
/// Client for fetching public transparency dashboard data.
/// All endpoints are public - no authentication required.

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter/foundation.dart';
import '../../config/web3_config.dart';

/// Dashboard overview model
class DashboardOverview {
  final int totalAttestations;
  final int totalResolutions;
  final double averageResolutionTimeHours;
  final int categoriesTracked;
  final int panchayatsActive;
  final DateTime? lastUpdated;
  final NetworkInfo network;

  DashboardOverview({
    required this.totalAttestations,
    required this.totalResolutions,
    required this.averageResolutionTimeHours,
    required this.categoriesTracked,
    required this.panchayatsActive,
    this.lastUpdated,
    required this.network,
  });

  factory DashboardOverview.fromJson(Map<String, dynamic> json) {
    final summary = json['summary'] ?? json;
    return DashboardOverview(
      totalAttestations: summary['totalAttestations'] ?? 0,
      totalResolutions: summary['totalResolutions'] ?? summary['activePanchayats'] ?? 0,
      averageResolutionTimeHours: (summary['averageResolutionTimeHours'] ?? summary['averageResolutionTime'] ?? 0).toDouble(),
      categoriesTracked: summary['categoriesTracked'] ?? 0,
      panchayatsActive: summary['panchayatsActive'] ?? summary['activePanchayats'] ?? 0,
      lastUpdated: json['lastUpdated'] != null 
          ? DateTime.fromMillisecondsSinceEpoch(json['lastUpdated']) 
          : null,
      network: NetworkInfo.fromJson(json['network'] ?? {}),
    );
  }
}

/// Network info model
class NetworkInfo {
  final String name;
  final int chainId;
  final String easContract;

  NetworkInfo({
    required this.name,
    required this.chainId,
    required this.easContract,
  });

  factory NetworkInfo.fromJson(Map<String, dynamic> json) {
    return NetworkInfo(
      name: json['name'] ?? 'Unknown',
      chainId: json['chainId'] ?? 0,
      easContract: json['easContract'] ?? '',
    );
  }
}

/// Category statistics model
class CategoryStats {
  final String name;
  final int count;
  final double percentage;

  CategoryStats({
    required this.name,
    required this.count,
    required this.percentage,
  });

  factory CategoryStats.fromJson(Map<String, dynamic> json) {
    return CategoryStats(
      name: json['name'] ?? 'Unknown',
      count: json['count'] ?? 0,
      percentage: (json['percentage'] ?? 0).toDouble(),
    );
  }
}

/// Panchayat ranking model
class PanchayatRanking {
  final String panchayatId;
  final int totalResolutions;
  final double averageResolutionTimeHours;
  final int efficiencyScore;
  final int rank;

  PanchayatRanking({
    required this.panchayatId,
    required this.totalResolutions,
    required this.averageResolutionTimeHours,
    required this.efficiencyScore,
    this.rank = 0,
  });

  factory PanchayatRanking.fromJson(Map<String, dynamic> json, [int index = 0]) {
    return PanchayatRanking(
      panchayatId: json['panchayatId'] ?? '',
      totalResolutions: json['totalResolutions'] ?? 0,
      averageResolutionTimeHours: (json['averageResolutionTimeHours'] ?? 0).toDouble(),
      efficiencyScore: json['efficiencyScore'] ?? 0,
      rank: index + 1,
    );
  }

  /// Get efficiency grade (A-F)
  String get grade {
    if (efficiencyScore >= 90) return 'A+';
    if (efficiencyScore >= 80) return 'A';
    if (efficiencyScore >= 70) return 'B';
    if (efficiencyScore >= 60) return 'C';
    if (efficiencyScore >= 50) return 'D';
    return 'F';
  }

  /// Get grade color
  int get gradeColor {
    if (efficiencyScore >= 90) return 0xFF4CAF50; // Green
    if (efficiencyScore >= 80) return 0xFF8BC34A; // Light Green
    if (efficiencyScore >= 70) return 0xFFFFC107; // Amber
    if (efficiencyScore >= 60) return 0xFFFF9800; // Orange
    if (efficiencyScore >= 50) return 0xFFFF5722; // Deep Orange
    return 0xFFF44336; // Red
  }
}

/// Daily trend point model
class TrendPoint {
  final DateTime date;
  final int count;

  TrendPoint({
    required this.date,
    required this.count,
  });

  factory TrendPoint.fromJson(Map<String, dynamic> json) {
    return TrendPoint(
      date: DateTime.parse(json['date']),
      count: json['count'] ?? 0,
    );
  }
}

/// Trend data model
class TrendData {
  final List<TrendPoint> points;
  final int total;
  final double average;

  TrendData({
    required this.points,
    required this.total,
    required this.average,
  });

  factory TrendData.fromJson(Map<String, dynamic> json) {
    final trend = json['trend'] as List<dynamic>? ?? [];
    return TrendData(
      points: trend.map((p) => TrendPoint.fromJson(p)).toList(),
      total: json['total'] ?? 0,
      average: (json['average'] ?? 0).toDouble(),
    );
  }
}

/// Recent attestation model (anonymized)
class RecentAttestation {
  final String uid;
  final String? category;
  final String? panchayatId;
  final double? resolutionTimeHours;
  final DateTime timestamp;
  final String? officerHash;

  RecentAttestation({
    required this.uid,
    this.category,
    this.panchayatId,
    this.resolutionTimeHours,
    required this.timestamp,
    this.officerHash,
  });

  factory RecentAttestation.fromJson(Map<String, dynamic> json) {
    return RecentAttestation(
      uid: json['uid'] ?? '',
      category: json['category'],
      panchayatId: json['panchayatId'],
      resolutionTimeHours: json['resolutionTimeHours']?.toDouble(),
      timestamp: json['timestamp'] != null 
          ? DateTime.fromMillisecondsSinceEpoch(json['timestamp']) 
          : DateTime.now(),
      officerHash: json['officerHash'],
    );
  }

  /// Get short UID for display
  String get shortUid {
    if (uid.length > 16) {
      return '${uid.substring(0, 8)}...${uid.substring(uid.length - 8)}';
    }
    return uid;
  }
}

/// Aggregate statistics model
class AggregateStats {
  final DashboardOverview overview;
  final List<CategoryStats> topCategories;
  final List<PanchayatRanking> topPanchayats;
  final TrendData weeklyTrend;
  final DateTime generatedAt;

  AggregateStats({
    required this.overview,
    required this.topCategories,
    required this.topPanchayats,
    required this.weeklyTrend,
    required this.generatedAt,
  });

  factory AggregateStats.fromJson(Map<String, dynamic> json) {
    return AggregateStats(
      overview: DashboardOverview.fromJson({'summary': json['overview']}),
      topCategories: (json['topCategories'] as List<dynamic>? ?? [])
          .map((c) => CategoryStats.fromJson(c))
          .toList(),
      topPanchayats: (json['topPanchayats'] as List<dynamic>? ?? [])
          .asMap()
          .entries
          .map((e) => PanchayatRanking.fromJson(e.value, e.key))
          .toList(),
      weeklyTrend: TrendData.fromJson(json['weeklyTrend'] ?? {}),
      generatedAt: json['generatedAt'] != null 
          ? DateTime.fromMillisecondsSinceEpoch(json['generatedAt']) 
          : DateTime.now(),
    );
  }
}

/// Dashboard Service
class DashboardService {
  final http.Client _client;
  static const Duration _timeout = Duration(seconds: 15);
  
  DashboardService({
    http.Client? client,
  }) : _client = client ?? http.Client();

  String get _baseUrl => Web3Config.attestationServiceUrl;
  
  // Debug getter to expose the base URL
  String get debugBaseUrl => _baseUrl;
  
  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    if (Web3Config.apiKey.isNotEmpty) 'x-api-key': Web3Config.apiKey,
  };

  /// Get dashboard overview
  Future<DashboardOverview> getOverview() async {
    try {
      debugPrint('DashboardService: Getting overview from $_baseUrl/dashboard/overview');
      final response = await _client.get(
        Uri.parse('$_baseUrl/dashboard/overview'),
        headers: _headers,
      ).timeout(_timeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return DashboardOverview.fromJson(data['data']);
      } else {
        throw DashboardException('Failed to get dashboard overview');
      }
    } catch (e) {
      if (e is DashboardException) rethrow;
      throw DashboardException('Network error: $e');
    }
  }

  /// Get attestations by category
  Future<List<CategoryStats>> getCategoryStats() async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/dashboard/categories'),
        headers: _headers,
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final categories = data['data']['categories'] as List<dynamic>? ?? [];
        return categories.map((c) => CategoryStats.fromJson(c)).toList();
      } else {
        throw DashboardException('Failed to get category stats');
      }
    } catch (e) {
      if (e is DashboardException) rethrow;
      throw DashboardException('Network error: $e');
    }
  }

  /// Get panchayat rankings
  Future<List<PanchayatRanking>> getPanchayatRankings() async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/dashboard/panchayats'),
        headers: _headers,
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final rankings = data['data']['rankings'] as List<dynamic>? ?? [];
        return rankings.asMap().entries
            .map((e) => PanchayatRanking.fromJson(e.value, e.key))
            .toList();
      } else {
        throw DashboardException('Failed to get panchayat rankings');
      }
    } catch (e) {
      if (e is DashboardException) rethrow;
      throw DashboardException('Network error: $e');
    }
  }

  /// Get daily trend
  Future<TrendData> getDailyTrend({int days = 30}) async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/dashboard/trend?days=$days'),
        headers: _headers,
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return TrendData.fromJson(data['data']);
      } else {
        throw DashboardException('Failed to get daily trend');
      }
    } catch (e) {
      if (e is DashboardException) rethrow;
      throw DashboardException('Network error: $e');
    }
  }

  /// Get recent attestations
  Future<List<RecentAttestation>> getRecentAttestations({int limit = 10}) async {
    try {
      debugPrint('DashboardService: Getting recent from $_baseUrl/dashboard/recent?limit=$limit');
      final response = await _client.get(
        Uri.parse('$_baseUrl/dashboard/recent?limit=$limit'),
        headers: _headers,
      ).timeout(_timeout);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final attestations = data['data']['attestations'] as List<dynamic>? ?? [];
        return attestations.map((a) => RecentAttestation.fromJson(a)).toList();
      } else {
        throw DashboardException('Failed to get recent attestations');
      }
    } catch (e) {
      debugPrint('DashboardService getRecentAttestations ERROR: $e');
      if (e is DashboardException) rethrow;
      throw DashboardException('Network error: $e');
    }
  }

  /// Get aggregate statistics
  Future<AggregateStats> getAggregateStats() async {
    try {
      debugPrint('DashboardService: Getting stats from $_baseUrl/dashboard/stats');
      final response = await _client.get(
        Uri.parse('$_baseUrl/dashboard/stats'),
        headers: _headers,
      ).timeout(_timeout);

      debugPrint('DashboardService: Got response ${response.statusCode}');
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return AggregateStats.fromJson(data['data']);
      } else {
        throw DashboardException('Failed to get aggregate stats: ${response.statusCode}');
      }
    } catch (e) {
      debugPrint('DashboardService ERROR: $e');
      if (e is DashboardException) rethrow;
      throw DashboardException('Network error: $e');
    }
  }

  /// Export metrics for archival
  Future<Map<String, dynamic>> exportMetrics() async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/dashboard/export'),
        headers: _headers,
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['data'] ?? {};
      } else {
        throw DashboardException('Failed to export metrics');
      }
    } catch (e) {
      if (e is DashboardException) rethrow;
      throw DashboardException('Network error: $e');
    }
  }

  /// Health check
  Future<bool> healthCheck() async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/dashboard/health'),
        headers: _headers,
      );
      return response.statusCode == 200;
    } catch (e) {
      debugPrint('Dashboard health check failed: $e');
      return false;
    }
  }

  void dispose() {
    _client.close();
  }
}

/// Dashboard exception
class DashboardException implements Exception {
  final String message;
  DashboardException(this.message);
  
  @override
  String toString() => 'DashboardException: $message';
}

File: backend/src/services/dashboardService.js
/**
 * Dashboard Service - Transparency & Analytics
 * 
 * Provides public transparency dashboard data:
 * - Attestation statistics
 * - Resolution metrics
 * - Officer performance
 * - Panchayat rankings
 */

const { ethers } = require('ethers');
const { config } = require('../config');
const logger = require('../utils/logger');
const easService = require('./easService');

class DashboardService {
  constructor() {
    this.provider = null;
    this.initialized = false;
    
    // In-memory metrics cache
    this.metricsCache = {
      totalAttestations: 0,
      totalResolutions: 0,
      averageResolutionTime: 0,
      attestationsByCategory: {},
      attestationsByPanchayat: {},
      dailyAttestations: [],
      monthlyTrends: [],
      topPerformers: [],
      lastUpdated: null,
    };
    
    // Resolution tracking
    this.resolutions = [];
  }

  /**
   * Initialize the dashboard service
   */
  async initialize() {
    if (this.initialized) {
      return;
    }

    try {
      this.provider = new ethers.JsonRpcProvider(config.blockchain.rpcUrl);
      this.initialized = true;
      logger.info('Dashboard service initialized successfully');
    } catch (error) {
      logger.error('Failed to initialize dashboard service', { error: error.message });
      throw error;
    }
  }

  /**
   * Record a new attestation for metrics
   * @param {Object} attestationData - Attestation details
   */
  recordAttestation(attestationData) {
    const {
      uid,
      issueId,
      category,
      panchayatId,
      officerId,
      resolutionTimeHours,
      timestamp = Date.now(),
    } = attestationData;

    // Update totals
    this.metricsCache.totalAttestations++;
    this.metricsCache.totalResolutions++;

    // Update category stats
    if (category) {
      this.metricsCache.attestationsByCategory[category] = 
        (this.metricsCache.attestationsByCategory[category] || 0) + 1;
    }

    // Update panchayat stats
    if (panchayatId) {
      if (!this.metricsCache.attestationsByPanchayat[panchayatId]) {
        this.metricsCache.attestationsByPanchayat[panchayatId] = {
          total: 0,
          averageTime: 0,
          resolutionTimes: [],
        };
      }
      const pStats = this.metricsCache.attestationsByPanchayat[panchayatId];
      pStats.total++;
      if (resolutionTimeHours) {
        pStats.resolutionTimes.push(resolutionTimeHours);
        pStats.averageTime = pStats.resolutionTimes.reduce((a, b) => a + b, 0) / 
                             pStats.resolutionTimes.length;
      }
    }

    // Track resolution
    this.resolutions.push({
      uid,
      issueId,
      category,
      panchayatId,
      officerId,
      resolutionTimeHours,
      timestamp,
    });

    // Update average resolution time
    const timesWithData = this.resolutions.filter(r => r.resolutionTimeHours);
    if (timesWithData.length > 0) {
      this.metricsCache.averageResolutionTime = 
        timesWithData.reduce((sum, r) => sum + r.resolutionTimeHours, 0) / timesWithData.length;
    }

    // Update daily stats
    this.updateDailyStats(timestamp);

    this.metricsCache.lastUpdated = Date.now();
    
    logger.debug('Attestation recorded for metrics', { uid, category, panchayatId });
  }

  /**
   * Update daily attestation statistics
   */
  updateDailyStats(timestamp) {
    const date = new Date(timestamp).toISOString().split('T')[0];
    const existing = this.metricsCache.dailyAttestations.find(d => d.date === date);
    
    if (existing) {
      existing.count++;
    } else {
      this.metricsCache.dailyAttestations.push({ date, count: 1 });
      // Keep only last 30 days
      if (this.metricsCache.dailyAttestations.length > 30) {
        this.metricsCache.dailyAttestations.shift();
      }
    }
  }

  /**
   * Get public dashboard overview
   */
  async getOverview() {
    await this.initialize();

    return {
      summary: {
        totalAttestations: this.metricsCache.totalAttestations,
        totalResolutions: this.metricsCache.totalResolutions,
        averageResolutionTimeHours: Math.round(this.metricsCache.averageResolutionTime * 10) / 10,
        categoriesTracked: Object.keys(this.metricsCache.attestationsByCategory).length,
        panchayatsActive: Object.keys(this.metricsCache.attestationsByPanchayat).length,
      },
      lastUpdated: this.metricsCache.lastUpdated,
      network: {
        name: config.blockchain.network,
        chainId: config.blockchain.chainId,
        easContract: config.blockchain.easContract,
      },
    };
  }

  /**
   * Get attestations by category
   */
  async getAttestationsByCategory() {
    const categories = this.metricsCache.attestationsByCategory;
    const total = Object.values(categories).reduce((sum, count) => sum + count, 0);

    return {
      categories: Object.entries(categories).map(([name, count]) => ({
        name,
        count,
        percentage: total > 0 ? Math.round((count / total) * 100) : 0,
      })).sort((a, b) => b.count - a.count),
      total,
    };
  }

  /**
   * Get panchayat performance rankings
   */
  async getPanchayatRankings() {
    const panchayats = this.metricsCache.attestationsByPanchayat;

    return {
      rankings: Object.entries(panchayats).map(([id, stats]) => ({
        panchayatId: id,
        totalResolutions: stats.total,
        averageResolutionTimeHours: Math.round(stats.averageTime * 10) / 10,
        efficiencyScore: this.calculateEfficiencyScore(stats),
      })).sort((a, b) => b.efficiencyScore - a.efficiencyScore),
    };
  }

  /**
   * Calculate efficiency score (0-100)
   */
  calculateEfficiencyScore(stats) {
    // Score based on volume and speed
    const volumeScore = Math.min(stats.total / 10, 50); // Max 50 points for volume
    const speedScore = stats.averageTime > 0 
      ? Math.max(0, 50 - (stats.averageTime / 2)) // Faster = more points, max 50
      : 25; // Default if no time data
    
    return Math.round(volumeScore + speedScore);
  }

  /**
   * Get daily attestation trend
   * @param {number} days - Number of days to return
   */
  async getDailyTrend(days = 30) {
    const data = this.metricsCache.dailyAttestations.slice(-days);
    
    // Fill in missing days with 0
    const filledData = [];
    const today = new Date();
    
    for (let i = days - 1; i >= 0; i--) {
      const date = new Date(today);
      date.setDate(date.getDate() - i);
      const dateStr = date.toISOString().split('T')[0];
      
      const existing = data.find(d => d.date === dateStr);
      filledData.push({
        date: dateStr,
        count: existing ? existing.count : 0,
      });
    }

    return {
      trend: filledData,
      total: filledData.reduce((sum, d) => sum + d.count, 0),
      average: Math.round(filledData.reduce((sum, d) => sum + d.count, 0) / days * 10) / 10,
    };
  }

  /**
   * Get recent attestations (public, anonymized)
   * @param {number} limit - Number of recent attestations
   */
  async getRecentAttestations(limit = 10) {
    const recent = this.resolutions
      .slice(-limit)
      .reverse()
      .map(r => ({
        uid: r.uid,
        category: r.category,
        panchayatId: r.panchayatId,
        resolutionTimeHours: r.resolutionTimeHours,
        timestamp: r.timestamp,
        // Anonymize officer ID
        officerHash: r.officerId ? 
          ethers.keccak256(ethers.toUtf8Bytes(r.officerId)).slice(0, 10) : null,
      }));

    return {
      attestations: recent,
      total: this.resolutions.length,
    };
  }

  /**
   * Get aggregate statistics
   */
  async getAggregateStats() {
    const categoryStats = await this.getAttestationsByCategory();
    const panchayatStats = await this.getPanchayatRankings();
    const dailyTrend = await this.getDailyTrend(7);

    return {
      overview: {
        totalAttestations: this.metricsCache.totalAttestations,
        averageResolutionTime: this.metricsCache.averageResolutionTime,
        activePanchayats: panchayatStats.rankings.length,
        categoriesTracked: categoryStats.categories.length,
      },
      topCategories: categoryStats.categories.slice(0, 5),
      topPanchayats: panchayatStats.rankings.slice(0, 5),
      weeklyTrend: dailyTrend,
      generatedAt: Date.now(),
    };
  }

  /**
   * Get verification statistics
   */
  async getVerificationStats() {
    // Track verification attempts (would be populated by verify endpoint)
    return {
      totalVerifications: 0,
      successfulVerifications: 0,
      failedVerifications: 0,
      verificationRate: 100,
      lastVerification: null,
    };
  }

  /**
   * Export metrics for IPFS archival
   */
  async exportMetricsForArchival() {
    const overview = await this.getOverview();
    const categories = await this.getAttestationsByCategory();
    const rankings = await this.getPanchayatRankings();
    const trend = await this.getDailyTrend(30);

    return {
      version: '1.0',
      exportedAt: Date.now(),
      network: config.blockchain.network,
      metrics: {
        overview: overview.summary,
        categories: categories.categories,
        panchayatRankings: rankings.rankings,
        dailyTrend: trend.trend,
      },
      signature: null, // Could be signed by service wallet
    };
  }

  /**
   * Health check for dashboard service
   */
  async healthCheck() {
    return {
      status: 'healthy',
      initialized: this.initialized,
      metricsCount: this.metricsCache.totalAttestations,
      lastUpdated: this.metricsCache.lastUpdated,
      cacheSize: {
        resolutions: this.resolutions.length,
        categories: Object.keys(this.metricsCache.attestationsByCategory).length,
        panchayats: Object.keys(this.metricsCache.attestationsByPanchayat).length,
      },
    };
  }
}

// Export singleton instance
module.exports = new DashboardService();
